packages/components/image.tsx

```diff
@@ -0,0 +1,122 @@
+import React from "react";
+import { connect, Head } from "frontity";
+import { Connect, Package } from "frontity/types";
+import useInView from "@frontity/hooks/use-in-view";
+
+// Hides any image rendered by this component that is not
+// inside a <noscript> when JS is disabled.
+const noJsStyles = `
+  :not(noscript) > .frontity-lazy-image {
+    display: none;
+  }
+`;
+
+// Finds all the images rendered by this component and maps
+// their `data-src` and `data-srcset` attributes to `src` and `srcset`
+// when the browser doesn't support Proxy or IntersectionObserver.
+const noProxyScript = `
+  if (typeof window !== "undefined" && (!("Proxy" in window) || !("IntersectionObserver" in window))) {
+    document.addEventListener("DOMContentLoaded", () => {
+      const images = document.querySelectorAll("img.frontity-lazy-image");
+      for (image of images) {
+        image.setAttribute("src", image.getAttribute("data-src"));
+        image.setAttribute("srcset", image.getAttribute("data-srcset"));
+        image.removeAttribute("data-src");
+        image.removeAttribute("data-srcset");
+        image.removeAttribute("style");
+      }
+    });
+  }
+`;
+
+interface Props {
+  src?: string;
+  srcSet?: string;
+  alt?: string;
+  className?: string;
+  loading?: "auto" | "lazy" | "eager";
+}
+
+type Image = React.FC<Connect<Package, Props>>;
+
+interface Attributes extends Props {
+  "data-src"?: string;
+  "data-srcset"?: string;
+  style?: { visibility: "hidden" };
+}
+
+type NoScriptImage = React.FC<Attributes>;
+
+interface ChangeAttributes {
+  (props: Attributes);
+}
+
+const changeAttributes: ChangeAttributes = attributes => {
+  attributes.src = attributes["data-src"];
+  attributes.srcSet = attributes["data-srcset"];
+  delete attributes["data-src"];
+  delete attributes["data-srcset"];
+  delete attributes["style"];
+};
+
+const NoScriptImage: NoScriptImage = props => {
+  const attributes = { ...props };
+  changeAttributes(attributes);
+  return (
+    <noscript>
+      <img {...attributes} />
+    </noscript>
+  );
+};
+
+const Image: Image = ({ src, srcSet, alt, loading, className }) => {
+  const attributes: Attributes = {
+    alt,
+    "data-src": src,
+    "data-srcset": srcSet,
+    className: "frontity-lazy-image".concat(className ? ` ${className}` : ""),
+    loading: loading || "auto",
+    style: { visibility: "hidden" }
+  };
+
+  if (typeof window !== "undefined") {
+    if (typeof (HTMLImageElement as any).prototype.loading !== "undefined") {
+      changeAttributes(attributes);
+      return <img {...attributes} />;
```

Comment:
Is `loading` here `"auto"` by default? Is that the correct default parameter? (instead of `"lazy"`):
![Screen Shot 2019-06-18 at 16 18 12](https://user-images.githubusercontent.com/3305402/59692230-c67a8500-91e4-11e9-9261-3cafaa63602f.png)

Reply:
Yep, if not specified the `loading` attribute will have the `auto` value.

Reply:
And what's the point of not using `"lazy"` by default?

I mean... for example, native lazy load won't work right now in Mars:
https://github.com/frontity/frontity/blob/51afabdbe02bdf590d42a9afb0b7d5380ad4ea57/packages/mars-theme/src/components/featured-media.js#L25-L29

Or maybe I'm not getting what `"auto"` means...

Reply:
The post only said that the decision is taken by the browser. How the browser decides, I have no clue. I just used `auto` as default because of the specification, but maybe we want to change it in the future. It won't be a breaking change until native lazy load is available.

Reply:
Also keep in mind that this is an `Image` component, not a `LazyImage` component. Lazy load is just one of the features.

Reply:
Oh, I see.

If that's the case maybe we should use the `loading` prop to opt-out of the IO implementation as well if the comp receives `"eager"`, shouldn't we?

Not that I think we should have that now, just that it would seem more consistent with how the component is used.

Reply:
That's right. That crossed my mind earlier but I forgot it once I was fighting with the implementation 😅

Reply:
I've made a fix. Let me know what you think.

Reply:
Changed some things with the next steps in the implementation in mind.